### PR TITLE
Add W3C tracecontext propagation and propagate full 128 bit B3 headers

### PR DESCRIPTION
**Pull request by bwoebi**
### Description

This PR foreshadows the work on 128 bit propagation to the backend and in datadog headers, to avoid special logic implementation for w3c headers.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

**Review by morrisonlevi — COMMENTED** [review]
I'd like to see the parsing and such done in a way that it can become a `component/` so that it can be independently tested without PHP at all. This is due to the fact that writing unit tests for C bits of PHP is a bad experience. Most of the time, but not all of the time, I also think separating it out as a component ends up being cleaner and nicer if done that way:

- Unit testing is done in the component, done entirely in the implementation language (not PHP).
- Integration/end-to-end tests is done by running through PHP programs.

Each one doesn't have to be its own module, maybe we have a `component/propagation-parsers` or something that holds all of them.

What do you think?

**Comment by bwoebi** [comment]
Given that some of that code is using HashTables, no (the propagated tags).
You could move it into zai/ and pass it a big struct of things to fill data into.
But I feel like separating the inputs and outputs from the code here would probably make the code messier.

The only advantage would be a more direct testability, but it's possible to come quite close to unit tests by using `DDTrace\consume_distributed_tracing_headers()` which basically exposes this parsing function directly to userland.

Ultimately, writing tests in PHP code also feels a lot more accessible to me, than using catch2. I don't like reading these, I don't like writing these.

So, I'd rather not.

File: tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+$spans = dd_trace_serialize_closed_spans();
+var_dump(isset($spans[0]["meta"]["_dd.propagation_error"]));
+
+echo 'Done.' . PHP_EOL;
+
+?>
+--EXPECTF--
+b3: 12345678901234567890123456789012-6543210987654321-d
+traceparent: 00-12345678901234567890123456789012-6543210987654321-01
+tracestate: dd=o:phpt-test;s:2;t.test:qvalue;unknown1:val;unknown2:1,foo=bar:;=,baz=qux
```
Comment (pierotibou) [inline tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt:48]: Could we handle more use cases? Special characters, invalid values...?

**Review by pierotibou — APPROVED** [review]
LGTM. I'm mainly questionning the default values for header propagation conf. Also could you add some utests on invalid scenarios, special chars? 

Finally, please next time, to make things easier for me don't squeeze in other changes (like the traceid one :prettyplease:).

File: ext/configuration.h [hunk 2]
```diff
@@ -99,8 +99,11 @@ extern bool runtime_config_first_init;
     CONFIG(SET_LOWERCASE, DD_TRACE_HEADER_TAGS, "")                                                            \
     CONFIG(INT, DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH, "512")                                                     \
     CONFIG(BOOL, DD_TRACE_PROPAGATE_SERVICE, "false")                                                          \
-    CONFIG(SET, DD_PROPAGATION_STYLE_EXTRACT, "Datadog,B3,B3 single header")                                   \
-    CONFIG(SET, DD_PROPAGATION_STYLE_INJECT, "Datadog")                                                        \
+    CALIAS(SET_LOWERCASE, DD_TRACE_PROPAGATION_STYLE_EXTRACT, "Tracecontext,Datadog,B3,B3 single header",      \
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/configuration.h:102]: That'd be a BC break - preserving the current behaviour. At least for extraction it doesn't make any harm to accept more, as long as tracestate and datadog are prioritized.

File: ext/ddtrace.c [hunk 3]
```diff
@@ -2265,18 +2276,139 @@ void ddtrace_read_distributed_tracing_ids(bool (*read_header)(zai_string_view, c
...
+                if (last_comma && ++commas == 30) {
+                    --persist;
+                    break;
+                }
+            }
+            ZSTR_LEN(DDTRACE_G(tracestate)) = persist - ZSTR_VAL(DDTRACE_G(tracestate));
+            zend_string_release(tracestate);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/ddtrace.c:2404]: That function grew over time and likely could be put into its own file at some point :-D

File: ext/ddtrace.c [hunk 4]
```diff
@@ -2265,18 +2276,139 @@ void ddtrace_read_distributed_tracing_ids(bool (*read_header)(zai_string_view, c
...
+                .low = ddtrace_parse_hex_span_id_str(start + 16, 16)
+            };
+
+            // parent-id
+            start = end + 1;
+            end = strchr(start, '-');
+            if (!end || end - start != 16) {
+                break;
+            }
+            DDTRACE_G(distributed_parent_trace_id) = ddtrace_parse_hex_span_id_str(start, 16);
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/ddtrace.c:2314]: It's the (span) id of the parent trace, that's why it was called that way :-D

File: ext/ddtrace.c [hunk 5]
```diff
@@ -2265,18 +2276,139 @@ void ddtrace_read_distributed_tracing_ids(bool (*read_header)(zai_string_view, c
...
+
+        if (read_header(ZAI_STRL_VIEW("X_TRACESTATE"), "tracestate", &tracestate, data)) {
+            bool last_comma = true;
+            DDTRACE_G(tracestate) = zend_string_alloc(ZSTR_LEN(tracestate), 0);
+            char *persist = ZSTR_VAL(DDTRACE_G(tracestate));
+            zend_hash_clean(&DDTRACE_G(tracestate_unknown_dd_keys));
+            int commas = 0;
+            for (char *ptr = ZSTR_VAL(tracestate), *end = ptr + ZSTR_LEN(tracestate); ptr < end; ++ptr) {
+                // dd member
+                if (last_comma && ptr + 2 < end && ptr[0] == 'd' && ptr[1] == 'd' && (ptr[2] == '=' || ptr[2] == '\t' || ptr[2] == ' ')) {
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/ddtrace.c:2332]: The document says `List-members may have leading or trailing optional whitespace (0x09 (horizontal tab) and 0x20 (space)), so be sure to exclude the whitespace when searching for the dd list-member.`, i.e. need to handle spaces after `dd`. Unless I misinterpret that?

File: ext/ddtrace.c [hunk 6]
```diff
@@ -2218,16 +2228,17 @@ void ddtrace_read_distributed_tracing_ids(bool (*read_header)(zai_string_view, c
...
+        DDTRACE_G(distributed_trace_id) = (ddtrace_trace_id){ .low = ddtrace_parse_userland_span_id(&trace_zv) };
         zend_string_release(trace_id_str);
     } else if (parse_b3 && read_header(ZAI_STRL_VIEW("X_B3_TRACEID"), "b3-traceid", &trace_id_str, data)) {
-        zval trace_zv;
-        ZVAL_STR(&trace_zv, trace_id_str);
-        DDTRACE_G(distributed_trace_id) = ddtrace_parse_hex_span_id(&trace_zv);
+        DDTRACE_G(distributed_trace_id) = (ddtrace_trace_id){
+            .high = ZSTR_LEN(trace_id_str) > 16 ? ddtrace_parse_hex_span_id_str(ZSTR_VAL(trace_id_str), MIN(16, ZSTR_LEN(trace_id_str) - 16)) : 0,
+            .low = ddtrace_parse_hex_span_id_str(ZSTR_VAL(trace_id_str) + MAX(0, (ssize_t)ZSTR_LEN(trace_id_str) - 16), MIN(16, ZSTR_LEN(trace_id_str))),
+        };
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/ddtrace.c:2237]: Makes sense

File: tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt (hunk 1 above)
Comment (bwoebi) [reply tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt:48]: I don't think like we need particularly well defined behaviour for invalid values as long as it doesn't have adverse side-effects.
What test inputs exactly would you propose?

File: ext/ddtrace.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (pierotibou) [reply ext/ddtrace.c:2314]: From my viewpoint, a trace has no parent. A span can have one but not a trace. But anyway, nevermind, as not related

File: tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt (hunk 1 above)
Comment (pierotibou) [reply tests/ext/integrations/curl/distributed_tracing_curl_tracestate.phpt:48]: I was thinking about adding the invalid characters that should be replaced in origin or propagated tags for instance. But that will be difficult with a test where you do both extraction and injection. Could we add a test where we set invalid data on a span and assert the propagated values?

**Comment by bwoebi** [comment]
@pierotibou Applied your comments :-)

**Comment by pierotibou** [comment]
Thanks a lot @bwoebi

**Comment by bwoebi** [comment]
@pierotibou I had to update the docker build to add mbstring extension to make tests pass - still mergeable?

**Pull request merged**